File: fastapi/app/services/redis_service.py
import os
import json
import redis
from threading import Thread
from datetime import datetime
from typing import Dict, Any, Optional

# Canais Redis
CHANNELS = {
    "NODE_TASK_UPDATES": "node:task_updates",
    "FASTAPI_TASK_UPDATES": "fastapi:task_updates",
    "MODEL_SELECTION": "model:selection",
    "QUERY_RESULTS": "query:results",
}

# Configuração do Redis
REDIS_HOST = os.getenv("REDIS_HOST", "localhost")
REDIS_PORT = int(os.getenv("REDIS_PORT", 6379))
REDIS_PASSWORD = os.getenv("REDIS_PASSWORD", None)
REDIS_TLS = os.getenv("REDIS_TLS", "false").lower() == "true"

# Cache local para armazenar atualizações de tarefas
task_cache = {}

# Cliente Redis
redis_client = redis.Redis(
    host=REDIS_HOST,
    port=REDIS_PORT,
    password=REDIS_PASSWORD,
    ssl=REDIS_TLS,
    decode_responses=True
)

# Processar mensagens em thread separada
def redis_listener():
    pubsub = redis_client.pubsub()
    pubsub.subscribe(CHANNELS["NODE_TASK_UPDATES"], CHANNELS["QUERY_RESULTS"])
    
    for message in pubsub.listen():
        if message["type"] == "message":
            try:
                data = json.loads(message["data"])
                request_id = data.get("requestId")
                
                if request_id:
                    # Atualizar cache local
                    if request_id not in task_cache:
                        task_cache[request_id] = {"updates": [], "results": None}
                    
                    channel = message["channel"]
                    
                    if channel == CHANNELS["NODE_TASK_UPDATES"]:
                        task_cache[request_id]["updates"].append(data)
                        print(f"Atualização via Redis: {request_id}")
                    
                    elif channel == CHANNELS["QUERY_RESULTS"]:
                        task_cache[request_id]["results"] = data.get("results")
                        print(f"Resultados recebidos para: {request_id}")
                        
                        # Limpar cache após um tempo ou implementar LRU cache
                        # TODO: Implementar limpeza de cache
                        
            except Exception as e:
                print(f"Erro ao processar mensagem Redis: {e}")

# Publicar seleção de modelo
def publish_model_selection(request_id: str, model: str, query: str):
    redis_client.publish(
        CHANNELS["MODEL_SELECTION"],
        json.dumps({
            "requestId": request_id,
            "model": model,
            "query": query,
            "timestamp": datetime.now().isoformat()
        })
    )

# Publicar atualizações de tarefas
def publish_task_update(request_id: str, update_type: str, payload: Any):
    redis_client.publish(
        CHANNELS["FASTAPI_TASK_UPDATES"],
        json.dumps({
            "requestId": request_id,
            "type": update_type,
            "payload": payload,
            "timestamp": datetime.now().isoformat()
        })
    )

# Obter atualizações de tarefas do cache
def get_task_updates(request_id: str) -> Dict[str, Any]:
    if request_id in task_cache:
        return task_cache[request_id]
    return {"updates": [], "results": None}

# Obter resultados de consulta
def get_query_results(request_id: str) -> Optional[Dict[str, Any]]:
    if request_id in task_cache and task_cache[request_id].get("results"):
        return task_cache[request_id]["results"]
    return None

# Iniciar thread de escuta
listener_thread = None

def start_redis_listener():
    global listener_thread
    if listener_thread is None or not listener_thread.is_alive():
        listener_thread = Thread(target=redis_listener, daemon=True)
        listener_thread.start()
        print("Listener Redis iniciado com sucesso") 

File: buscador_inteligente/src/services/redis-service.ts
import Redis from "ioredis";
import { EventEmitter } from "events";

// Canais Redis
export const CHANNELS = {
  NODE_TASK_UPDATES: "node:task_updates",
  FASTAPI_TASK_UPDATES: "fastapi:task_updates",
  MODEL_SELECTION: "model:selection",
  QUERY_RESULTS: "query:results",
  NCM_REQUEST: "ncm:request",
  NCM_RESPONSE: "ncm:response",
};

// Configuração do Redis
// Usando a porta 6378 conforme configurado no sistema CMEX
const redisConfig = {
  host: process.env.REDIS_HOST || "localhost",
  port: Number(process.env.REDIS_PORT) || 6379, // Será 6378 quando definido no .env
  password: process.env.REDIS_PASSWORD,
  tls: process.env.REDIS_TLS === "true" ? {} : undefined,
};

// Clientes Redis
export const publisher = new Redis(redisConfig);
export const subscriber = new Redis(redisConfig);

// Inicializar assinaturas
export function initializeRedisSubscriptions(eventEmitter: EventEmitter) {
  console.log(
    `Assinado nos canais Redis com sucesso em ${redisConfig.host}:${redisConfig.port}`
  );

  // Assinar aos canais relevantes usando Promise em vez de callback
  subscriber
    .subscribe(
      CHANNELS.FASTAPI_TASK_UPDATES,
      CHANNELS.MODEL_SELECTION,
      CHANNELS.NCM_REQUEST
    )
    .then(() => {
      // Mensagem já exibida acima
    })
    .catch((err) => {
      console.error("Erro ao assinar canais Redis:", err);
    });

  // Manipular mensagens recebidas
  subscriber.on("message", (channel: string, message: string) => {
    try {
      const data = JSON.parse(message);

      if (channel === CHANNELS.FASTAPI_TASK_UPDATES && data.requestId) {
        console.log(
          `Atualização recebida via Redis para tarefa ${data.requestId}`
        );

        // Emitir evento para clientes conectados
        eventEmitter.emit(`progress-${data.requestId}`, {
          type: data.type || "progress",
          data: data.payload,
        });
      } else if (channel === CHANNELS.MODEL_SELECTION && data.requestId) {
        console.log(
          `Seleção de modelo recebida: ${data.model} para ${data.requestId}`
        );

        // Processar a seleção do modelo
        processModelSelection(data.requestId, data.model, data.query);
      } else if (channel === CHANNELS.NCM_REQUEST && data.requestId) {
        console.log(`Requisição NCM recebida: ${data.requestId}`);

        // Responder à solicitação de teste de integração
        processNcmRequest(data.requestId, data.query);
      }
    } catch (error) {
      console.error("Erro ao processar mensagem Redis:", error);
    }
  });
}

// Publicar atualização de tarefa
export function publishTaskUpdate(
  requestId: string,
  updateType: string,
  payload: any
) {
  publisher.publish(
    CHANNELS.NODE_TASK_UPDATES,
    JSON.stringify({
      requestId,
      type: updateType,
      payload,
      timestamp: new Date().toISOString(),
    })
  );
}

// Publicar resultados de consulta
export function publishQueryResults(requestId: string, results: any) {
  publisher.publish(
    CHANNELS.QUERY_RESULTS,
    JSON.stringify({
      requestId,
      results,
      timestamp: new Date().toISOString(),
    })
  );
}

// Processar seleção de modelo
async function processModelSelection(
  requestId: string,
  modelName: string,
  query: string
) {
  // Implementar lógica para usar o modelo selecionado
  // Esta função deve iniciar o processamento com o modelo correto
  console.log(
    `Iniciando processamento para ${requestId} com modelo ${modelName}`
  );

  // Aqui você deve implementar a lógica para iniciar o processamento
  // com o modelo selecionado, integrando com seu sistema existente

  // Exemplo:
  // const result = await startProcessingWithModel(requestId, query, modelName);
  // publishQueryResults(requestId, result);
}

// Processar requisição NCM (para teste de integração)
async function processNcmRequest(requestId: string, query: string) {
  console.log(`Processando requisição NCM: ${requestId} - Query: ${query}`);

  // Simular processamento
  setTimeout(() => {
    // Enviar resposta
    publisher.publish(
      CHANNELS.NCM_RESPONSE,
      JSON.stringify({
        requestId,
        response: `Resposta do Node.js para query: ${query}`,
        timestamp: new Date().toISOString(),
      })
    );
    console.log(`Resposta enviada para ${requestId}`);
  }, 1000); // Atraso de 1 segundo para simular processamento
}

// Função para encerrar conexões Redis
export function closeRedisConnections() {
  try {
    if (publisher && publisher.status === "ready") {
      publisher
        .quit()
        .catch((err) => console.error("Erro ao fechar publisher Redis:", err));
    }

    if (subscriber && subscriber.status === "ready") {
      subscriber
        .quit()
        .catch((err) => console.error("Erro ao fechar subscriber Redis:", err));
    }

    console.log("Conexões Redis encerradas com sucesso");
  } catch (error) {
    console.error("Erro ao encerrar conexões Redis:", error);
  }
}
